src/python/pants/core/goals/init.py

```diff
@@ -0,0 +1,224 @@
+    return ret
+
+
+@rule
+async def edit_build_files(req: EditBuildFilesRequest) -> EditedBuildFiles:
+    ptgts_by_build_file = group_by_build_file(req.putative_targets)
+    existing_build_files_contents = await Get(DigestContents, PathGlobs(ptgts_by_build_file.keys()))
+    existing_build_files_contents_by_path = {
+        ebfc.path: ebfc.content for ebfc in existing_build_files_contents
+    }
+
+    def make_content(bf_path: str, pts: Iterable[PutativeTarget]) -> FileContent:
+        existing_content = existing_build_files_contents_by_path.get(bf_path)
+        new_content = ([] if existing_content is None else [existing_content.decode()]) + [
+            pt.generate_build_file_stanza() for pt in pts
+        ]
+        new_content_bytes = "\n\n".join(new_content).encode()
+        return FileContent(bf_path, new_content_bytes)
+
+    new_digest = await Get(
+        Digest,
+        CreateDigest([make_content(path, ptgts) for path, ptgts in ptgts_by_build_file.items()]),
+    )
+
+    updated = set(existing_build_files_contents_by_path.keys())
+    created = set(ptgts_by_build_file.keys()) - updated
+    return EditedBuildFiles(new_digest, tuple(sorted(created)), tuple(sorted(updated)))
+
+
+@goal_rule
+async def init(
+    init_subsystem: InitSubsystem,
+    console: Console,
+    workspace: Workspace,
+    union_membership: UnionMembership,
+) -> Init:
+    putative_target_request_types = union_membership[PutativeTargetsRequest]
+    putative_target_reqs = [req_type() for req_type in putative_target_request_types]
+    putative_targets_results = await MultiGet(
+        Get(PutativeTargets, PutativeTargetsRequest, req) for req in putative_target_reqs
+    )
+    all_putative_targets = PutativeTargets.merge(putative_targets_results)
+    all_existing_tgts = await Get(Targets, AddressSpecs([DescendantAddresses("")]))
+    addr_to_existing_tgt = {
+        f"{tgt.address.spec_path}:{tgt.address.target_name}": tgt for tgt in all_existing_tgts
+    }
+
+    conflicting_ptgts: List[PutativeTarget] = []
+    nonconflicting_ptgts: List[PutativeTarget] = []
+
+    for ptgt in all_putative_targets:
+        (
+            conflicting_ptgts if ptgt.address_spec in addr_to_existing_tgt else nonconflicting_ptgts
+        ).append(ptgt)
+
+    conflicting_ptgts_by_build_file = group_by_build_file(conflicting_ptgts)
+    nonconflicting_ptgts_by_build_file = group_by_build_file(nonconflicting_ptgts)
+
+    # Edit BUILD files for the putative targets whose addresses don't conflict with
+    # those of existing targets.
+    edited_build_files = await Get(
+        EditedBuildFiles,
+        EditBuildFilesRequest(
+            PutativeTargets(
+                ptgt
+                for ptgt in all_putative_targets
+                if ptgt.address_spec not in addr_to_existing_tgt
+            )
+        ),
+    )
+    updated_build_files = set(edited_build_files.updated_paths)
+
+    workspace.write_digest(edited_build_files.digest)
```

Comment:
We only want to call this if there were actual changes made, for performance reasons and to avoid things like invalidating IDE buffers. Maybe add a property to `EditedBuildFiles`? This is how we do it for `fmt.py`:

https://github.com/pantsbuild/pants/blob/ed293aada5091d3888235ebedc119f99672349c5/src/python/pants/core/goals/fmt.py#L73-L75

Reply:
This is now only ever run if there were changes.